### PR TITLE
Iostats command implementation

### PR DESCRIPTION
**Pull request by utkarshmani1997**
This commit adds the command iostats in istgtcontrol to get
the iostats from istgt.

On branch iostats-command-implemantation
Changes to be committed:
	modified:   istgt_lu.h
	modified:   istgt_lu_ctl.c
	modified:   istgt_lu_disk.c
	modified:   istgtcontrol.c

1. Why is this change necessary ?
-  There is no such command in istgtcontrol to get these details
   so this command adds the additional functionality to get iostats

2. How to verify this change ?
-  Build the binary and run `sudo istgtcontrol iostats` and you
   will get the output something like given below.
```
$ sudo istgtcontrol iostats
{ "iqn": "iqn.2017-08.OpenEBS.cstor:vol1", "writes": "0", "reads": "0", "totalwritebytes": "0", "totalreadbytes": "0", "size": "10737418240" } DONE IOSTATS command
```

3. What side effects does this change have ?
-  none

Special notes for the reviewer :
- This does not displays the complets iostats but only displays the
  available stats.

Signed-off-by: Utkarsh Mani Tripathi <utkarshmani1997@gmail.com>

### CODE REVIEW

File: src/istgt_lu_ctl.c [hunk 1]
```diff
@@ -2845,7 +2845,37 @@ _verb_stat SCSIstat_rslt[SCSI_ARYSZ] = { {0,0,0} };
...
+                if (lu == NULL)
+                        continue;
+                for (j = 0; j< lu->maxlun; j++) { 
+                	spec = (ISTGT_LU_DISK *) lu->lun[j].spec;
+			if (spec == NULL)
+				continue;
+                	istgt_uctl_snprintf(uctl, "%s  IQN=%s Blockcount=%lu Blocklength=%lu Writes=%lu Reads=%lu TotalReadBytes=%lu TotalWriteBytes=%lu Size=%lu\n", uctl->cmd, lu->name, spec->blockcnt, spec->blocklen, spec->writes, spec->reads, spec->readbytes, spec->writebytes, spec->size);
+                	rc = istgt_uctl_writeline(uctl);
+                }
+                if (rc != UCTL_CMD_OK)
```
Comment (vishnuitta) [inline src/istgt_lu_ctl.c]: you should be writing error on the socket

File: src/istgt_lu_disk.c [hunk 2]
```diff
@@ -114,9 +114,14 @@ size_t rcommon_cmd_mempool_count = RCOMMON_CMD_MEMPOOL_ENTRIES;
 		if(cmd_write) {\
 			rcomm_cmd->opcode = ZVOL_OPCODE_WRITE;\
 			rcomm_cmd->iovcnt = cmd->iobufindx+1;\
+                        __sync_add_and_fetch(&spec->writes, 1);\
+			__sync_add_and_fetch(&spec->writebytes, nbytes);\
 		} else {\
 			rcomm_cmd->opcode = ZVOL_OPCODE_READ;\
 			rcomm_cmd->iovcnt = 0;\
+			printf("got read\n");	\
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: debug message!

File: src/istgt_lu_ctl.c [hunk 3]
```diff
@@ -3057,6 +3087,7 @@ static ISTGT_UCTL_CMD_TABLE istgt_uctl_cmd_table[] =
 	{ "RSV", istgt_uctl_cmd_rsv},
 	{ "QUE", istgt_uctl_cmd_que},
 	{ "STATS", istgt_uctl_cmd_stats},
+        { "IOSTATS", istgt_uctl_cmd_iostats},
```
Comment (vishnuitta) [inline src/istgt_lu_ctl.c]: intendation

File: src/istgtcontrol.c [hunk 4]
```diff
@@ -1349,9 +1389,10 @@ static EXEC_TABLE exec_table[] =
 	{"LOG", exec_log, 0, 0 },
 	{"RSV", exec_rsv, 0, 0 },
 	{"QUE", exec_que, 0, 0 },
-	{"STATS", exec_stats, 0, 0 },
-	{"SET", exec_set, 0, 1},
-	{"MAXTIME", exec_maxtime, 0, 0},
+        {"STATS", exec_stats, 0, 0 },
+        {"SET", exec_set, 0, 1},
+       	{"IOSTATS", exec_iostats, 0, 0 },
```
Comment (vishnuitta) [inline src/istgtcontrol.c]: intendation

File: src/istgtcontrol.c [hunk 5]
```diff
@@ -1066,6 +1066,46 @@ exec_dump(UCTL_Ptr uctl)
 	return UCTL_CMD_OK;
 }
 
+static int
+exec_iostats(UCTL_Ptr uctl)
+{
+        const char *delim = ARGS_DELIM;
+        char *arg, buf[1000];
```
Comment (vishnuitta) [inline src/istgtcontrol.c]: unused variables buf and var

File: src/istgt_lu_ctl.c [hunk 6]
```diff
@@ -2845,7 +2845,37 @@ _verb_stat SCSIstat_rslt[SCSI_ARYSZ] = { {0,0,0} };
...
+        ISTGT_Ptr istgt = uctl->istgt;
+        for (i = 0; i < MAX_LOGICAL_UNIT; i++) {
+                lu = (ISTGT_LU_Ptr *) istgt->logical_unit[i];
+                if (lu == NULL)
+                        continue;
+                for (j = 0; j< lu->maxlun; j++) { 
+                	spec = (ISTGT_LU_DISK *) lu->lun[j].spec;
+			if (spec == NULL)
+				continue;
+                	istgt_uctl_snprintf(uctl, "%s  IQN=%s Blockcount=%lu Blocklength=%lu Writes=%lu Reads=%lu TotalReadBytes=%lu TotalWriteBytes=%lu Size=%lu\n", uctl->cmd, lu->name, spec->blockcnt, spec->blocklen, spec->writes, spec->reads, spec->readbytes, spec->writebytes, spec->size);
```
Comment (vishnuitta) [inline src/istgt_lu_ctl.c]: why such a long line?

File: src/istgtcontrol.c [hunk 7]
```diff
@@ -1066,6 +1066,46 @@ exec_dump(UCTL_Ptr uctl)
...
+              break;
+           }
+           if (uctl->iqn != NULL) {
+		printf("%s\n", arg);
+ 	   } else {
+		printf("%s\n", arg);
+	   }
+        }
+        if (strcmp(result, "OK") != 0) {
+		if (is_err_req_auth(uctl, arg))
```
Comment (vishnuitta) [inline src/istgtcontrol.c]: this is not required

Comment (utkarshmani1997) [reply src/istgtcontrol.c]: ok

File: src/istgt_lu_ctl.c (hunk 1 above)
Comment (utkarshmani1997) [reply src/istgt_lu_ctl.c]: ok

File: test_istgt.sh [hunk 8]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-
+set -x
```
Comment (mynktl) [inline test_istgt.sh]: don't enable debug for this script.

File: src/istgtcontrol.c [hunk 9]
```diff
@@ -1066,6 +1066,44 @@ exec_dump(UCTL_Ptr uctl)
...
+		arg = trim_string(uctl->recvbuf);
+		result = strsepq(&arg, delim);
+		strupr(result);
+		if (strcmp(result, uctl->cmd) !=0){
+			break;
+		}
+		if (uctl->iqn != NULL) {
+			printf("%s\n", arg);
+		} else {
+			printf("%s\n", arg);
```
Comment (mynktl) [inline src/istgtcontrol.c]: here, you are printing arg in both the cases. What is the difference here?
should we handle IQN number?

Comment (utkarshmani1997) [reply src/istgtcontrol.c]: this is the case with the every command in the istgtcontrol. Anyway i'm removing it.

File: src/istgt_lu_disk.c [hunk 10]
```diff
@@ -124,9 +124,13 @@ size_t rcommon_cmd_mempool_count = RCOMMON_CMD_MEMPOOL_ENTRIES;
 		if (cmd_write) {						\
 			rcomm_cmd->opcode = ZVOL_OPCODE_WRITE;		\
 			rcomm_cmd->iovcnt = cmd->iobufindx+1;		\
+      __sync_add_and_fetch(&spec->writes, 1);\
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: intendation.. in below lines also

**Comment by utkarshmani1997** [comment]
will raise new PR for this, so closing this now.

**Pull request closed without merging**